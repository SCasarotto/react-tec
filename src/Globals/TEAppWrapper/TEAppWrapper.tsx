import React from 'react'
import { BrowserRouter, useLocation } from 'react-router-dom'
import { Manager } from 'react-popper'

import { useScrollTop } from '../../Utility/useScrollTop'
import { TEThemeProvider, TEThemeProviderProps } from '../TEThemeContext'
import { TEPopupProvider } from '../TEPopupContext'

export interface TEAppWrapperProps extends TEThemeProviderProps {}
const TEAppWrapperContent: React.FC<TEAppWrapperProps> = (props) => {
	const { theme, globalStyles, children } = props
	const location = useLocation()
	useScrollTop(location)

	return (
		<TEThemeProvider theme={theme} globalStyles={globalStyles}>
			<TEPopupProvider>
				<Manager>{children}</Manager>
			</TEPopupProvider>
		</TEThemeProvider>
	)
}

/**
 *
 * TEAppWrapper is meant to be the single top level wrapping element you need. Having said that, it does contain a number of things. `TEAppWrapper` contain the following:
 *
 * ```
 * <BrowserRouter>
 * 	<TEScrollToTop>
 * 		<TEThemeProvider theme={theme} globalStyles={globalStyles}>
 *			<TEPopupProvider>{children}</TEPopupProvider>
 *		</TEThemeProvider>
 * 	</TEScrollToTop>
 * </BrowserRouter>
 * ```
 *
 * Here `BrowserRouter` comes from [react-router-dom](https://reacttraining.com/react-router/web/api/BrowserRouter). For all other components, please view thier given documentation page.
 *
 * As for props, the only props `TEAppWrapper` accepts are those that `TEThemeProvider` accepts.
 *
 * `TEAppWrapper` is implemented like the following in this project:
 *
 * ```
 * const Main = () => (
 * 	<TEAppWrapper globalStyles={globalStyles}>
 *		<App />
 * 	</TEAppWrapper>
 * )
 * ReactDOM.render(<Main />, document.getElementById('root'))
 * ```
 *
 */
export const TEAppWrapper: React.FC<TEAppWrapperProps> = (props) => {
	return (
		<BrowserRouter>
			<TEAppWrapperContent {...props} />
		</BrowserRouter>
	)
}
