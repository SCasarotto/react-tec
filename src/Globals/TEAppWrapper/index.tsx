import React from 'react'
import {
	BrowserRouter,
	RouteComponentProps,
	useLocation,
} from 'react-router-dom'

import { useScrollTop } from '../../Utility/useScrollTop'
import { TEThemeProvider, TEThemeProviderProps } from '../TEThemeContext'
import { TEPopupProvider, TEPopupProviderProps } from '../TEPopupContext'
interface TEAppWrapperProps
	extends TEThemeProviderProps,
		TEPopupProviderProps,
		RouteComponentProps {}

const TEAppWrapperContent: React.FC<TEAppWrapperProps> = (props) => {
	const { theme, globalStyles, children } = props
	const location = useLocation()
	useScrollTop(location)

	return (
		<TEThemeProvider theme={theme} globalStyles={globalStyles}>
			<TEPopupProvider>{children}</TEPopupProvider>
		</TEThemeProvider>
	)
}

export const TEAppWrapper: React.FC<TEAppWrapperProps> = (props) => {
	return (
		<BrowserRouter>
			<TEAppWrapperContent {...props} />
		</BrowserRouter>
	)
}
