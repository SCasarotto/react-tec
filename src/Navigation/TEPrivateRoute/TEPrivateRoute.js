import React from 'react'
import { Route, Redirect } from 'react-router-dom'

const TEPrivateRoute = (props) => {
	const {
		isAuthenticated,
		hasPermissions,
		authPath,
		accessDeniedPath,
		component: Component,
		...rest
	} = props
	return (
		<Route
			{...rest}
			render={(props) => {
				if (isAuthenticated && hasPermissions) {
					return Component ? <Component {...props} /> : <Route {...rest} />
				} else if (!isAuthenticated) {
					return <Redirect to={authPath} />
				} else if (!hasPermissions) {
					return <Redirect to={accessDeniedPath} />
				} else {
					//Can't Happen
					return null
				}
			}}
		/>
	)
}

TEPrivateRoute.defaultProps = {
	hasPermissions: true,
	authPath: '/signin',
	accessDeniedPath: '/403',
}

export default TEPrivateRoute
