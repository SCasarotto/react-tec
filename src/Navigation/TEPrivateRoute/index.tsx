import React from 'react'
import { Route, Redirect, RouteProps, RouteComponentProps } from 'react-router-dom'

interface TEPrivateRouteAuthCheck {
	check(props: RouteComponentProps<any>): boolean
	path: string
}
export interface TEPrivateRouteProps extends RouteProps {
	authChecks?: TEPrivateRouteAuthCheck[]
}
export const TEPrivateRoute: React.FC<TEPrivateRouteProps> = (props) => {
	const { authChecks, component: Component, ...rest } = props

	return (
		<Route
			{...rest}
			render={(props) => {
				if (authChecks) {
					for (let i = 0; i < authChecks.length; i++) {
						const { check, path } = authChecks[i]
						if (!check(props)) {
							return (
								<Redirect
									to={{
										pathname: path,
										state: { fromPath: rest.path },
									}}
								/>
							)
						}
					}
				}

				return Component ? <Component {...props} /> : <Route {...rest} />
			}}
		/>
	)
}
