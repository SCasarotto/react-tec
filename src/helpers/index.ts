export * from './colorHelpers'
export * from './manipulateRowData'
