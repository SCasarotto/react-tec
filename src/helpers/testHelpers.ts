export const mockOnScroll = () => {
	const { scrollTo } = window
	beforeAll(() => {
		// Replace with custom value
		window.scrollTo = jest.fn()
	})
	afterAll(() => {
		// Restore original
		window.scrollTo = scrollTo
	})
}
