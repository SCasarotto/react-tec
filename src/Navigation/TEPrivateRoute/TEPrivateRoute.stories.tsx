import React, { useState } from 'react'
import { Meta } from '@storybook/react'

import { TEPrivateRoute } from './TEPrivateRoute'
import { Link, Route, Switch } from 'react-router-dom'
import { TEButton } from '../../Form/TEButton'

export default {
	title: 'Navigation/TEPrivateRoute',
	component: TEPrivateRoute,
	argTypes: {},
} as Meta

const Base = () => <p>Base Route</p>
const Authenticated = () => <p>Authenticated Route</p>
const AuthenticatedAdmin = () => <p>AuthenticatedAdmin Route</p>
const AuthenticatedAdminDynamicData = (props: any) => (
	<p>AuthenticatedAdmin Route - data: {props.match.params.someData}</p>
)
const AuthenticatedFailed = () => <p>AuthenticatedFailed Route</p>
const AuthenticatedAdminFailed = () => <p>AuthenticatedAdminFailed Route</p>
const NotFound = () => <p>NotFound Route</p>

export const Default = () => {
	const [auth, setAuth] = useState(false)
	const [authAdmin, setAuthAdmin] = useState(false)

	return (
		<>
			<ul>
				<li>
					<Link to='/TEPrivateRouteExample'>Base</Link>
				</li>
				<li>
					<Link to='/TEPrivateRouteExample/authenticated'>Authenticated</Link>
				</li>
				<li>
					<Link to='/TEPrivateRouteExample/authenticatedAdmin'>AuthenticatedAdmin</Link>
				</li>
				<li>
					<Link to='/TEPrivateRouteExample/authenticatedAdmin/someDynamicData'>
						AuthenticatedAdmin & someDynamicData
					</Link>
				</li>
			</ul>
			<TEButton onClick={() => setAuth(!auth)}>
				Toggle Authenticated To {auth ? 'False' : 'True'}
			</TEButton>
			<TEButton onClick={() => setAuthAdmin(!authAdmin)}>
				Toggle AuthenticatedAdmin To {authAdmin ? 'False' : 'True'}
			</TEButton>
			<Switch>
				<Route path='/TEPrivateRouteExample/' component={Base} exact />
				<TEPrivateRoute
					path='/TEPrivateRouteExample/authenticated'
					component={Authenticated}
					authChecks={[{ check: () => auth, path: '/TEPrivateRouteExample/sign-in' }]}
				/>
				<TEPrivateRoute
					path='/TEPrivateRouteExample/authenticatedAdmin/:someData'
					component={AuthenticatedAdminDynamicData}
					authChecks={[
						{ check: () => auth, path: '/TEPrivateRouteExample/sign-in' },
						{
							check: (props) =>
								authAdmin && props.match.params.someData === 'someDynamicData',
							path: '/TEPrivateRouteExample/403',
						},
					]}
				/>
				<TEPrivateRoute
					path='/TEPrivateRouteExample/authenticatedAdmin'
					component={AuthenticatedAdmin}
					authChecks={[
						{ check: () => auth, path: '/TEPrivateRouteExample/sign-in' },
						{
							check: () => authAdmin,
							path: '/TEPrivateRouteExample/403',
						},
					]}
				/>
				<Route path='/TEPrivateRouteExample/sign-in' component={AuthenticatedFailed} />
				<Route path='/TEPrivateRouteExample/403' component={AuthenticatedAdminFailed} />
				<Route component={NotFound} />
			</Switch>
		</>
	)
}
