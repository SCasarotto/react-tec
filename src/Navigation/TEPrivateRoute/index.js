import TEPrivateRoute from './TEPrivateRoute'

export default TEPrivateRoute
