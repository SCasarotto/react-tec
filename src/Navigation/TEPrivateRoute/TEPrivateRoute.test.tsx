import React from 'react'
import { render } from '@testing-library/react'
import { TEPrivateRoute } from './TEPrivateRoute'
import { TEAppWrapper } from './../../Globals/TEAppWrapper'
import { mockOnScroll } from '../../helpers/testHelpers'

describe('TEPrivateRoute', () => {
	mockOnScroll()
	it('matches snapshot', () => {
		const { asFragment } = render(
			<TEAppWrapper>
				<TEPrivateRoute />
			</TEAppWrapper>,
		)
		expect(asFragment()).toMatchSnapshot()
	})
})
