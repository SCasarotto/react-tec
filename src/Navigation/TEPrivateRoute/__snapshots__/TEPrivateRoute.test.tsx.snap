// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEPrivateRoute matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  position: fixed;
  width: 100%;
  height: 100%;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  background-color: rgba(0,0,0,0.5);
  z-index: 1000;
  overflow-y: auto;
  -webkit-transition: visibility 0.25s ease-in-out,opacity 0.25s ease-in-out;
  transition: visibility 0.25s ease-in-out,opacity 0.25s ease-in-out;
  opacity: 0;
  visibility: hidden;
}

.c2 {
  position: relative;
  background-color: #FFFFFF;
  border-radius: 20px;
  margin: 50px auto;
  max-width: 650px;
}

.c3 {
  position: relative;
  display: block;
  width: 100%;
  font-size: 18px;
  padding: 5px 20px;
  text-align: center;
  cursor: pointer;
  border-radius: 5px;
  box-sizing: border-box;
  margin: 0px;
  color: #5E5E5E;
  background-color: #FFFFFF;
  border: 1px solid #E5E5E5;
  -webkit-transition: color 0.2s ease-in,background-color 0.2s ease-in,border 0.2s ease-in;
  transition: color 0.2s ease-in,background-color 0.2s ease-in,border 0.2s ease-in;
}

.c3:hover,
.c3:active {
  color: #FFFFFF;
  background-color: #007CC3;
  border: 1px solid #007CC3;
}

.c3:disabled {
  cursor: not-allowed;
  color: #C5C5C5;
  border: 1px solid #F4F4F4;
}

.c1 .TEPopupContent {
  width: 90%;
  max-width: 600px;
  margin: 100px auto;
}

.c4 {
  border-top: 1px solid #007CC3;
  border-right: none;
  border-bottom: none;
  border-left: none;
  line-height: 2;
  border-top-left-radius: 0px;
  border-top-right-radius: 0px;
  border-bottom-left-radius: 20px;
  border-bottom-right-radius: 20px;
  -webkit-transition: border 0.2s ease-in,color 0.2s ease-in,background-color 0.2s ease-in;
  transition: border 0.2s ease-in,color 0.2s ease-in,background-color 0.2s ease-in;
}

.c4:active,
.c4:hover {
  color: #FFFFFF;
  background-color: #007CC3;
  border-top: 1px solid #007CC3;
  border-right: none;
  border-bottom: none;
  border-left: none;
}

@media (max-width:400px),(max-height:550px) {

}

<div
    class="c0 TEPopup c1 TEAlert "
  >
    <div
      class="c2 TEPopupContent"
    >
      <button
        class="c3 TEButton c4 TEAlertButton"
        type="button"
      >
        Okay
      </button>
    </div>
  </div>
  .c0 {
  position: fixed;
  width: 100%;
  height: 100%;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  background-color: rgba(0,0,0,0.5);
  z-index: 1000;
  overflow-y: auto;
  -webkit-transition: visibility 0.25s ease-in-out,opacity 0.25s ease-in-out;
  transition: visibility 0.25s ease-in-out,opacity 0.25s ease-in-out;
  opacity: 0;
  visibility: hidden;
}

.c2 {
  position: relative;
  background-color: #FFFFFF;
  border-radius: 20px;
  margin: 50px auto;
  max-width: 650px;
}

.c4 {
  position: relative;
  display: block;
  width: 100%;
  font-size: 18px;
  padding: 5px 20px;
  text-align: center;
  cursor: pointer;
  border-radius: 5px;
  box-sizing: border-box;
  margin: 0px;
  color: #5E5E5E;
  background-color: #FFFFFF;
  border: 1px solid #E5E5E5;
  -webkit-transition: color 0.2s ease-in,background-color 0.2s ease-in,border 0.2s ease-in;
  transition: color 0.2s ease-in,background-color 0.2s ease-in,border 0.2s ease-in;
}

.c4:hover,
.c4:active {
  color: #FFFFFF;
  background-color: #007CC3;
  border: 1px solid #007CC3;
}

.c4:disabled {
  cursor: not-allowed;
  color: #C5C5C5;
  border: 1px solid #F4F4F4;
}

.c1 .TEPopupContent {
  width: 90%;
  max-width: 600px;
  margin: 100px auto;
}

.c3 {
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  width: 50%;
  font-size: 18px;
  line-height: 2;
  border-top-left-radius: 0px;
  border-top-right-radius: 0px;
  border-bottom-left-radius: 20px;
  border-bottom-right-radius: 0px;
  border-top: 1px solid #007CC3;
  border-right: 1px solid #007CC3;
  border-bottom: none;
  border-left: none;
}

.c5:hover,
.c5:active {
  color: #FFFFFF;
  background-color: #007CC3;
  border-top: 1px solid #007CC3;
  border-right: 1px solid #007CC3;
  border-bottom: none;
  border-left: none;
}

.c6 {
  width: 50%;
  font-size: 18px;
  line-height: 2;
  border-top-left-radius: 0px;
  border-top-right-radius: 0px;
  border-bottom-left-radius: 0px;
  border-bottom-right-radius: 20px;
  border-top: 1px solid #007CC3;
  border-right: none;
  border-bottom: none;
  border-left: 1px solid #007CC3;
}

.c6:hover,
.c6:active {
  color: #FFFFFF;
  background-color: #007CC3;
  border-top: 1px solid #007CC3;
  border-right: none;
  border-bottom: none;
  border-left: 1px solid #007CC3;
}

@media (max-width:400px),(max-height:550px) {

}

<div
    class="c0 TEPopup c1 TEConfirm "
  >
    <div
      class="c2 TEPopupContent"
    >
      <div
        class="c3 TEConfirmButtonoContainer"
      >
        <button
          class="c4 TEButton c5 TEConfirmButton TEConfirmLeftButton"
          type="button"
        >
          Cancel
        </button>
        <button
          class="c4 TEButton c6 TEConfirmButton TEConfirmRightButton"
          type="button"
        />
      </div>
    </div>
  </div>
  .c0 {
  position: fixed;
  width: 100%;
  height: 100%;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  background-color: rgba(0,0,0,0.5);
  z-index: 1000;
  overflow-y: auto;
  -webkit-transition: visibility 0.25s ease-in-out,opacity 0.25s ease-in-out;
  transition: visibility 0.25s ease-in-out,opacity 0.25s ease-in-out;
  opacity: 0;
  visibility: hidden;
}

.c2 {
  position: relative;
  background-color: #FFFFFF;
  border-radius: 20px;
  margin: 50px auto;
  max-width: 650px;
}

.c3 {
  text-align: center;
}

.c5 {
  position: relative;
  display: inline-block;
  width: 80px;
  height: 80px;
}

.c6 {
  position: absolute;
  border-radius: 50%;
  overflow: hidden;
  border-top: 3px solid transparent;
  border-right: 1px solid transparent;
  border-bottom: 1px solid transparent;
  border-left: 1px solid transparent;
  left: 0%;
  top: 0%;
  width: 100%;
  height: 100%;
  border-top-color: #007cc3;
  -webkit-animation: TESpinnerLoading 1s linear infinite;
  animation: TESpinnerLoading 1s linear infinite;
}

.c7 {
  position: absolute;
  border-radius: 50%;
  overflow: hidden;
  border-top: 3px solid transparent;
  border-right: 1px solid transparent;
  border-bottom: 1px solid transparent;
  border-left: 1px solid transparent;
  left: 8.333333333333332%;
  top: 8.333333333333332%;
  width: 83.33333333333334%;
  height: 83.33333333333334%;
  border-top-color: #3397cf;
  -webkit-animation: TESpinnerLoading 1.125s linear infinite;
  animation: TESpinnerLoading 1.125s linear infinite;
}

.c8 {
  position: absolute;
  border-radius: 50%;
  overflow: hidden;
  border-top: 3px solid transparent;
  border-right: 1px solid transparent;
  border-bottom: 1px solid transparent;
  border-left: 1px solid transparent;
  left: 16.666666666666664%;
  top: 16.666666666666664%;
  width: 66.66666666666667%;
  height: 66.66666666666667%;
  border-top-color: #66b1db;
  -webkit-animation: TESpinnerLoading 1.2857142857142856s linear infinite;
  animation: TESpinnerLoading 1.2857142857142856s linear infinite;
}

.c9 {
  position: absolute;
  border-radius: 50%;
  overflow: hidden;
  border-top: 3px solid transparent;
  border-right: 1px solid transparent;
  border-bottom: 1px solid transparent;
  border-left: 1px solid transparent;
  left: 25%;
  top: 25%;
  width: 50%;
  height: 50%;
  border-top-color: #9acbe7;
  -webkit-animation: TESpinnerLoading 1.5s linear infinite;
  animation: TESpinnerLoading 1.5s linear infinite;
}

.c10 {
  position: absolute;
  border-radius: 50%;
  overflow: hidden;
  border-top: 3px solid transparent;
  border-right: 1px solid transparent;
  border-bottom: 1px solid transparent;
  border-left: 1px solid transparent;
  left: 33.33333333333333%;
  top: 33.33333333333333%;
  width: 33.333333333333336%;
  height: 33.333333333333336%;
  border-top-color: #cce5f3;
  -webkit-animation: TESpinnerLoading 1.7999999999999998s linear infinite;
  animation: TESpinnerLoading 1.7999999999999998s linear infinite;
}

.c11 {
  position: absolute;
  border-radius: 50%;
  overflow: hidden;
  border-top: 3px solid transparent;
  border-right: 1px solid transparent;
  border-bottom: 1px solid transparent;
  border-left: 1px solid transparent;
  left: 41.666666666666664%;
  top: 41.666666666666664%;
  width: 16.666666666666675%;
  height: 16.666666666666675%;
  border-top-color: #ffffff;
  -webkit-animation: TESpinnerLoading 2.2499999999999996s linear infinite;
  animation: TESpinnerLoading 2.2499999999999996s linear infinite;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 .TEPopupContent {
  background-color: rgba(0,0,0,0.5);
  padding: 60px;
  text-align: center;
}

.c4 {
  margin-bottom: 30px;
}

@media (max-width:400px),(max-height:550px) {

}

<div
    class="c0 TEPopup c1 TENetworkActivity "
  >
    <div
      class="c2 TEPopupContent"
    >
      <div
        class="c3 TESpinner c4 TENetworkActivitySpinner"
      >
        <div
          class="c5 TESpinnerContainer"
        >
          <div
            class="c6 TESpinnerRing TESpinnerRing1"
          />
          <div
            class="c7 TESpinnerRing TESpinnerRing2"
          />
          <div
            class="c8 TESpinnerRing TESpinnerRing3"
          />
          <div
            class="c9 TESpinnerRing TESpinnerRing4"
          />
          <div
            class="c10 TESpinnerRing TESpinnerRing5"
          />
          <div
            class="c11 TESpinnerRing TESpinnerRing6"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
